src/mpi/comm/comm_split_type.c

```diff
@@ -493,6 +627,63 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
     goto fn_exit;
 }
 
+#undef FUNCNAME
+#define FUNCNAME MPIR_Comm_split_type_network_topo
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+int MPIR_Comm_split_type_network_topo(MPIR_Comm * comm_ptr, int key,
+                                      MPIR_Info * info_ptr, MPIR_Comm ** newcomm_ptr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int info_args_are_equal, flag = 0;
+    char hintval[MPI_MAX_INFO_VAL + 1];
+
+    if (info_ptr) {
+        MPIR_Info_get_impl(info_ptr, NETWORK_INFO_KEY, MPI_MAX_INFO_VAL, hintval, &flag);
+    }
+
+    if (!flag) {
+        hintval[0] = '\0';
+    }
+
+    mpi_errno = compare_info_hint(hintval, comm_ptr, &info_args_are_equal);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /* if all processes do not have the same hintval, skip
+     * topology-aware comm split */
+    if (!info_args_are_equal)
+        goto use_null_comm;
+
+#ifdef HAVE_HWLOC
+    /* if our bindset is not valid, skip topology-aware comm split */
+    if (!MPIR_Process.bindset_is_valid)
+        goto use_null_comm;
+#endif
+
+#ifdef HAVE_NETLOC
+    if (!strncmp(hintval, ("switch_level:"), strlen("switch_level:"))
+        && *(hintval + strlen("switch_level:")) != '\0') {
+        int switch_level = atoi(hintval + strlen("switch_level:"));
+        mpi_errno = network_split_switch_level(comm_ptr, key, switch_level, newcomm_ptr);
+        goto fn_exit;
+    } else if (!strncmp(hintval, "subcomm_min_size:", strlen("subcomm_min_size:"))
+               && *(hintval + strlen("subcomm_min_size:")) != '\0') {
+        int size = atoi(hintval + strlen("subcomm_min_size:"));
+        mpi_errno = network_split_by_size(comm_ptr, key, size, newcomm_ptr);
+        goto fn_exit;
+    }
+#endif
+  use_null_comm:
```

Comment:
It might look better to set to `NULL` at the beginning of the function and then overwrite it later when conditions are met. Personally, all these `goto`s make me a little uncomfortable.